### ISSUE TITLE
Idea: Rate limit resource

### ISSUE THREAD

**Issue by Jeffail** [Jeffail/benthos#72]
It might be nice to have a rate limit field in the `resources` section of a Benthos config, where we might specify a maximum rate per interval for an arbitrary resource. This rate limit could then potentially be shared across inputs, processors or outputs that choose to use them.

For example, suppose we are using the `http` processor to hit external services, and in order to get a high throughput we have a large number of processing pipelines in a config something like this:

``` yaml
input:
  type: foo
buffer:
  type: memory
pipeline:
  threads: 40
  processors:
  - type: http
    http:
      max_parallel: 50
      url: http://fooresource.bar/process
```

This rate is fairly predictable, we can see that 40 threads each with a `max_parallel` of 50 would potentially create 2000 requests in parallel at any given time. We should be able to tune the parallelism, processor pipeline threads, and retries of the `http` processor in order to prevent stressing out the resource.

However, imagine we had a more complex pipeline where hitting this `http` resource is conditional on other factors such as the content of a message passing through. In this case it might be difficult to predict how often messages of a batch would be sent to this resource per interval per processing thread, and it's also possible that certain events may result in an odd distribution of these messages, therefore spiking the number of requests.

If our service is brittle and could potentially die as a result of being hit too often (as oppose to simply returning a 429) we might need some mechanism for rate limiting our requests in a way that is shared across pipeline processing threads and conditional processors of those pipelines. If we were able to specify a shared rate limit we could do something like this:

``` yaml
input:
  type: foo
buffer:
  type: memory
pipeline:
  threads: 40
  processors:
  - type: http
    http:
      max_parallel: 50
      rate_limit: foo_limiter
      url: http://fooresource.bar/process
resources:
  rate_limiters:
    foo_limiter:
      interval: 1s
      capacity: 1000
```

This would create a shared rate limit across all the `http` processors that specify the rate limiter `foo_limiter`, where the requests are limited to 1000 per second. It would then be easy to cap the usage of a particular resource even when your processing pipelines are complex. These rate limits could potentially be used by any `inputs`, `outputs`, `processors`, `caches`, etc where it seems sensible.

**Comment by Jeffail** [comment]
https://github.com/Jeffail/benthos/commit/b3db9b8c4cee52515b881eb84b88e3faaf54aa3d
